molnsutil/molnsutil.py

```diff
@@ -597,99 +649,234 @@ def save_state(self, name):
+        
         self.parameters = state['parameters']
         self.number_of_trajectories = state['number_of_trajectories']
         self.seed_base = state['seed_base']
         self.result_list = state['result_list']
         self.storage_mode = state['storage_mode']
+        #
+        self.reduced_results = state['reduced_results']
+        self.mapped_results = state['mapped_results']
+        self.number_of_results = state['number_of_results']
+        self.step1_complete = state['step1_complete']
+        self.step2_complete = state['step2_complete']
+        self.step3_complete = state['step3_complete']
+        self.mapper_fn = state['mapper_fn']
+        self.aggregator_fn  = state['aggregator_fn']
+        self.reducer_fn = state['reducer_fn']
+        #
+        if state['running_MapReduceTask'] is None:
+            self.running_MapReduceTask = None
+        else:
+            self.running_MapReduceTask = self.c.get_result(state['running_MapReduceTask'])
+        if state['running_SimulationTask'] is None:
+            self.running_SimulationTask = None
+        else:
+            self.running_SimulationTask = self.c.get_result(state['running_SimulationTask'])
 
-    #--------------------------
+    #-----------------------------------------------------------------------------------
     # MAIN FUNCTION
-    #--------------------------
-    def run(self, mapper, aggregator=None, reducer=None, number_of_trajectories=None, chunk_size=None, verbose=True, progress_bar=True, store_realizations=True, storage_mode="Shared", cache_results=False):
-        """ Main entry point """
-        if store_realizations:
-            if self.storage_mode is None:
-                if storage_mode != "Persistent" and storage_mode != "Shared":
-                    raise MolnsUtilException("Acceptable values for 'storage_mode' are 'Persistent' or 'Shared'")
-                self.storage_mode = storage_mode
-            elif self.storage_mode != storage_mode:
-                raise MolnsUtilException("Storage mode already set to {0}, can not mix storage modes".format(self.storage_mode))
-            # Do we have enough trajectores yet?
-            if number_of_trajectories is None and self.number_of_trajectories == 0:
-                raise MolnsUtilException("number_of_trajectories is zero")
-            # Run simulations
-            if self.number_of_trajectories < number_of_trajectories:
-                self.add_realizations( number_of_trajectories - self.number_of_trajectories, chunk_size=chunk_size, verbose=verbose, storage_mode=storage_mode)
+    #-----------------------------------------------------------------------------------
+    def run(self, mapper=None, aggregator=None, reducer=None, number_of_trajectories=None, chunk_size=None, verbose=True, progress_bar=True, store_realizations=True, storage_mode="Shared", cache_results=False):
```

Comment:
Should run use the identity mapper as a default? (to make it possible to get the same behavior as add_realization when we remove that)